lang/java/reef-common/src/main/java/org/apache/reef/driver/restart/DriverRestartManager.java

```diff
@@ -123,34 +129,32 @@ public synchronized void setPreviousEvaluatorIds(final Set<String> ids) {
   }
 
   /**
-   * @return the IDs of the Evaluators from a previous Driver that have checked in with the Driver
-   * in a restart situation.
-   */
-  public synchronized Set<String> getRecoveredEvaluatorIds() {
-    return Collections.unmodifiableSet(this.previousEvaluators);
-  }
-
-  /**
    * Indicate that this Driver has re-established the connection with one more Evaluator of a previous run.
+   * Calls the restart complete action if the latest evaluator is the last evaluator to recover.
    * @return true if the driver restart is completed.
    */
   public synchronized boolean onRecoverEvaluatorIsRestartComplete(final String evaluatorId) {
-    if (!this.previousEvaluators.contains(evaluatorId)) {
+    if (!this.previousEvaluators.containsKey(evaluatorId) ||
+        this.previousEvaluators.get(evaluatorId) == EvaluatorRestartState.NOT_RESTARTED_EVALUATOR) {
       final String errMsg = "Evaluator with evaluator ID " + evaluatorId + " not expected to be alive.";
       LOG.log(Level.SEVERE, errMsg);
       throw new DriverFatalRuntimeException(errMsg);
     }
 
-    if (!this.recoveredEvaluators.add(evaluatorId)) {
+    if (this.previousEvaluators.get(evaluatorId) != EvaluatorRestartState.EXPECTED) {
       LOG.log(Level.WARNING, "Evaluator with evaluator ID " + evaluatorId + " added to the set" +
           " of recovered evaluators more than once. Ignoring second add...");
+    } else {
+      setEvaluatorReported(evaluatorId);
     }
 
-    if (this.recoveredEvaluators.containsAll(this.previousEvaluators)) {
-      this.state = DriverRestartState.RestartCompleted;
+    for (final EvaluatorRestartState evaluatorRestartState : this.previousEvaluators.values()) {
```

Comment:
This could be a `private` method of its own. e.g. `haveAllExpectedEvaluatorsReported()`.